src/themis/secure_session_peer.c

```diff
@@ -45,6 +46,13 @@ themis_status_t secure_session_peer_init(secure_session_peer_t* peer,
         return THEMIS_INVALID_PARAMETER;
     }
 
+    /* This change prevents from using RSA keys in Secure Session,
+     * as they are currently not supported */
+    themis_key_kind_t key_kind = themis_get_asym_key_kind(sign_key, sign_key_len);
+    if (key_kind != THEMIS_KEY_EC_PRIVATE && key_kind != THEMIS_KEY_EC_PUBLIC) {
+        return THEMIS_INVALID_PARAMETER;
+    }
```

Comment:
Why is this check necessary, given that the key type is already checked in `secure_session_init`? Why is it either private or public?

Also, I'd put a similar comment _there_ in `secure_session_init` as well.

Reply:
Allow `THEMIS_KEY_EC_PUBLIC` since `secure_session_peer_init()` is called from `secure_session_proceed_client()`.
Allow `THEMIS_KEY_EC_PRIVATE` since `secure_session_peer_init()` is called from `secure_session_init()`.
Only check for private key in `secure_session_init()` would, theoretically, allow other peer to send non-EC key and cause errors. Theoretically because in current situation `secure_session_proceed_client()` doesn't even reach the call to `secure_session_peer_init()` if other side tries to use RSA key.
On the other hand, allowing public or private key in `secure_session_peer_init()` would allow creating session w/ public key and then fail while doing handshake. Skipping check for private/public key would break session creation / connection negotiation respectively.
Double check to make sure that user can neither create session with anything except private EC key nor the other peer can send unexpected key type.
I hope I answered your question :) Should I mention all of this (in compressed form) in the code?

Reply:
🤔

I've just reviewed the calls to secure_session_peer_init(). I think the checks are already performed, in a more... hardcoded manner.

1. secure_session_init() will get additional check added in this PR.

https://github.com/cossacklabs/themis/blob/86d9e705fc55c20b39d23d370110954c93a33d34/src/themis/secure_session.c#L109-L112

2. In secure_session_accept():

https://github.com/cossacklabs/themis/blob/86d9e705fc55c20b39d23d370110954c93a33d34/src/themis/secure_session.c#L418-L424

which already checks that `peer_sign_key` is EC public key:

https://github.com/cossacklabs/themis/blob/86d9e705fc55c20b39d23d370110954c93a33d34/src/themis/secure_session.c#L375-L379

3. In secure_session_proceed_client():

https://github.com/cossacklabs/themis/blob/86d9e705fc55c20b39d23d370110954c93a33d34/src/themis/secure_session.c#L614-L620

which also checks for EC public key beforehand:

https://github.com/cossacklabs/themis/blob/86d9e705fc55c20b39d23d370110954c93a33d34/src/themis/secure_session.c#L582-L584

It seems like adding additional checks here is redundant. `secure_session_peer_init()` is not public API, so we can be sure that those three callers are all there is to it.

Reply:
OK, so the check in `secure_session_peer_init()` will be removed